asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py

```diff
@@ -0,0 +1,926 @@
+            [
+                "installation_company_name",
+                "installation_company_mcs_number",
+                "certification_body",
+                "technology_type",
+            ]
+        )
+        .agg(
+            start_date=("commissioning_date", np.min),
+            end_date=("commissioning_date", np.max),
+        )
+        .unstack()
+    )
+
+    # Renaming columns to match those in the historical installers table e.g. air_source_hp_start_date
+    new_column_names = [
+        rename_columns(missing_installers.columns.get_level_values(1))[i]
+        + "_"
+        + missing_installers.columns.get_level_values(0)[i]
+        for i in range(len(missing_installers.columns))
+    ]
+    missing_installers.columns = new_column_names
+    missing_installers.reset_index(inplace=True)
+
+    # original_record is set to False as these are not originally in the installers table
+    missing_installers["original_record"] = [
+        False for i in range(len(missing_installers))
+    ]
+
+    # fuzzy match installer companies NOT in installer company data using companies house API to get address info
+    missing_installers[["full_address", "date_of_creation"]] = missing_installers.apply(
+        lambda x: match_companies_house(
+            x["installation_company_name"], companies_house_api_key
+        ),
+        axis=1,
+    )
+
+    # date_of_creation type change: str -> datetime
+    missing_installers["date_of_creation"] = missing_installers[
+        "date_of_creation"
+    ].apply(lambda x: datetime.strptime(x, "%Y-%m-%d") if not pd.isnull(x) else x)
+
+    # We only keep address info if the company creation date is before the first comissioning date (serves as a check)
+    start_date_cols = [col for col in missing_installers if "start_date" in col]
+    missing_installers["first_commissioning_date"] = missing_installers[
+        start_date_cols
+    ].min(axis=1)
+
+    missing_installers["full_address"] = missing_installers.apply(
+        lambda x: recompute_full_adress(
+            x["date_of_creation"], x["first_commissioning_date"], x["full_address"]
+        ),
+        axis=1,
+    )
+
+    # extract address_1 and postcode from full_address variable
+    missing_installers["address_1"] = missing_installers["full_address"].apply(
+        lambda x: x.split(",")[0] if x is not None else None
+    )
+    missing_installers["postcode"] = missing_installers["full_address"].apply(
+        lambda x: x.split(",")[-1].strip() if x is not None else None
+    )
+
+    # rename columns to match installers table
+    missing_installers.rename(
+        {
+            "installation_company_name": "company_name",
+            "installation_company_mcs_number": "mcs_certificate_number",
+        },
+        axis=1,
+        inplace=True,
+    )
+
+    # dropping unecessary variables
+    missing_installers.drop(
+        ["date_of_creation"],
+        axis=1,
+        inplace=True,
+    )
```

Comment:
For .rename and .drop functions, you can also use they keyword argument 'columns' instead of defining the axis. Doens't really matter but I'm always paranoid about mixing up the axes. :D 

```suggestion
    missing_installers.rename(columns=
        {
            "installation_company_name": "company_name",
            "installation_company_mcs_number": "mcs_certificate_number",
        }
        inplace=True,
    )

    # dropping unecessary variables
    missing_installers.drop(columns=
        ["date_of_creation"],
        inplace=True,
    )
```

Reply:
yep, using `columns` or `axis=1` is the same, but i'll change it :) it's definitely easier to understand with `columns` :)